### PR TITLE
feat: implement PromQL set op AND/UNLESS

### PR DESCRIPTION
**Pull request by waynexia**
I hereby agree to the terms of the [GreptimeDB CLA](https://gist.github.com/xtang/6378857777706e568c1949c7578592cc)

## What's changed and what's your intention?

implement `AND` and `UNLESS` for PromQL.

## Checklist

- [x]  I have written the necessary rustdoc comments.
- [x]  I have added the necessary unit tests and integration tests.

## Refer to a related PR or issue link (optional)
- #1042

### CODE REVIEW

File: src/promql/src/planner.rs [hunk 1]
```diff
@@ -1351,6 +1377,107 @@ impl PromPlanner {
...
+        right_tag_cols: Vec<String>,
+        op: TokenType,
+        modifier: &Option<BinModifier>,
+    ) -> Result<LogicalPlan> {
+        let mut left_tag_col_set = left_tag_cols.into_iter().collect::<HashSet<_>>();
+        let mut right_tag_col_set = right_tag_cols.into_iter().collect::<HashSet<_>>();
+
+        // apply modifier
+        if let Some(modifier) = modifier {
+            // one-to-many and many-to-one are not supported
```
Comment (killme2008) [inline src/promql/src/planner.rs:1394]: Will they be supported in the future?  If so, maybe change it to `TODO` item. If not, why?

File: src/promql/src/planner.rs [hunk 2]
```diff
@@ -1351,6 +1377,107 @@ impl PromPlanner {
...
+                    JoinType::LeftAnti,
+                    (join_keys.clone(), join_keys),
+                    None,
+                    true,
+                )
+                .context(DataFusionPlanningSnafu)?
+                .build()
+                .context(DataFusionPlanningSnafu),
+            token::T_LOR => {
+                // `OR` can not be expressed by `UNION` precisely.
```
Comment (killme2008) [inline src/promql/src/planner.rs:1470]: Maybe we can support the common case that uses `or` for the same schema?

**Review by killme2008 — COMMENTED** [review]
Looks Cooool To Me!

File: src/promql/src/planner.rs (hunk 1 above)
Comment (waynexia) [reply src/promql/src/planner.rs:1394]: This limitation is from PromQL. `group_left` and `group_right`:
>Grouping modifiers can only be used for [comparison](https://prometheus.io/docs/prometheus/latest/querying/operators/#comparison-binary-operators) and [arithmetic](https://prometheus.io/docs/prometheus/latest/querying/operators/#arithmetic-binary-operators). Operations as and, unless and or operations match with all possible entries in the right vector by default.

https://prometheus.io/docs/prometheus/latest/querying/operators/#group-modifiers

Comment (killme2008) [reply src/promql/src/planner.rs:1394]:
```
any-to-one and one-to-many matching are advanced use cases that should be carefully considered. Often a proper use of ignoring(<labels>) provides the desired outcome.
```

Looks like it supports but doesn't recommend using them.

Comment (waynexia) [reply src/promql/src/planner.rs:1394]: That's another context. For set operation:
>Grouping modifiers can only be used for [comparison](https://prometheus.io/docs/prometheus/latest/querying/operators/#comparison-binary-operators) and [arithmetic](https://prometheus.io/docs/prometheus/latest/querying/operators/#arithmetic-binary-operators).

File: tests/cases/standalone/common/promql/set_operation.result
Comment (yuanbohan) [inline tests/cases/standalone/common/promql/set_operation.result:1]: is it possible to include the original examples from https://prometheus.io/docs/prometheus/latest/querying/operators/#many-to-one-and-one-to-many-vector-matches?

File: src/promql/src/planner.rs (hunk 2 above)
Comment (waynexia) [reply src/promql/src/planner.rs:1470]: After some struggles I find it's not viable even for the same schemas 😢 Duplicate records also result to wrong data

File: tests/cases/standalone/common/promql/set_operation.result
Comment (yuanbohan) [reply tests/cases/standalone/common/promql/set_operation.result:1]: it's ok not including the set operation.

I can assure the expected results of the original example, but I'm not sure if your example is correct.

```
method_code:http_errors:rate5m{method="get", code="500"}  24
method_code:http_errors:rate5m{method="get", code="404"}  30
method_code:http_errors:rate5m{method="put", code="501"}  3
method_code:http_errors:rate5m{method="post", code="500"} 6
method_code:http_errors:rate5m{method="post", code="404"} 21

method:http_requests:rate5m{method="get"}  600
method:http_requests:rate5m{method="del"}  34
method:http_requests:rate5m{method="post"} 120
```

Comment (waynexia) [reply tests/cases/standalone/common/promql/set_operation.result:1]: I have put the original Prometheus case and result along with mine. E.g.:
```sql

-- eval instant at 50m http_requests{group="canary"} and http_requests{instance="0"}
-- 	http_requests{group="canary", instance="0", job="api-server"} 300
-- 	http_requests{group="canary", instance="0", job="app-server"} 700
-- SQLNESS SORT_RESULT 3 1
tql eval (3000, 3000, '1s') http_requests{g="canary"} and http_requests{instance="0"};

+---------------------+-----+----------+--------+-------+
| ts                  | job | instance | g      | val   |
+---------------------+-----+----------+--------+-------+
| 1970-01-01T00:50:00 | api | 0        | canary | 300.0 |
| 1970-01-01T00:50:00 | app | 0        | canary | 700.0 |
+---------------------+-----+----------+--------+-------+
```

**Review by yuanbohan — APPROVED** [review]

File: src/promql/src/planner.rs (hunk 2 above)
Comment (killme2008) [reply src/promql/src/planner.rs:1470]: Do we have any plan to support it in the future in other ways?

Comment (waynexia) [reply src/promql/src/planner.rs:1470]: I'm going to make a `UnionDistinctOn` plan for it

**Review by killme2008 — APPROVED** [review]
LGTM.

**Pull request merged**